### PR TITLE
improve rename

### PR DESCRIPTION
**Pull request by jimmyxian**
Improve rename so that it can refresh container immediately
#637 
Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @vieux @aluzzardi @abronan

**Comment by jimmyxian** [comment]
ping @vieux @aluzzardi @abronan

**Comment by vieux** [comment]
Thank you very much! 

can you please make sure all the commits are green in travis, it's much easier for us.
also, could you add an integration test ?

**Comment by aluzzardi** [comment]
Awesome, thanks!

@vieux I think @jimmyxian made this PR because his `rename` integration test didn't work in the first place.

@jimmyxian: Could you please extract the `rename` integration test from #637 and put it in this PR?

Thanks

**Comment by jimmyxian** [comment]
@vieux @aluzzardi Thks
I will move `rename` integration test from #637 and put it here.

**Comment by jimmyxian** [comment]
@vieux @aluzzardi 
Done to move.

File: api/handlers.go [hunk 1]
```diff
@@ -489,6 +489,34 @@ func postCommit(c *context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	if err := r.ParseForm(); err != nil {
+		httpError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// check new name whether avaliable
+	newName := r.Form.Get("name")
+	if conflictContainer := c.cluster.Container(newName); conflictContainer != nil {
```
Comment (aluzzardi) [inline api/handlers.go]: This check should be in `Cluster.Rename`

File: cluster/cluster.go [hunk 2]
```diff
@@ -43,4 +43,7 @@ type Cluster interface {
 	// FIXME: remove this method
 	// Return a random engine
 	RANDOMENGINE() (*Engine, error)
+
+	// Rename a container
+	Rename(container *Container, newName string) error
```
Comment (aluzzardi) [inline cluster/cluster.go]: This should be `RenameContainer` to be consistent with `CreateContainer`, `RemoveContainer`, ...

File: test/integration/api.bats [hunk 3]
```diff
@@ -352,9 +352,27 @@ function teardown() {
 	skip
 }
 
-# FIXME
 @test "docker rename" {
-	skip
+	start_docker 3
```
Comment (aluzzardi) [inline test/integration/api.bats]: Could you check conflict detection in the integration test?

You could create a container named `test-1` on node-1 and a container named `test-2` on node-2.

Then try to rename `test-2` into `test-1` and check that it fails.

Then try to rename `test-2` into `test-3` and check that it succeeds.

File: api/handlers.go [hunk 4]
```diff
@@ -489,6 +489,34 @@ func postCommit(c *context, w http.ResponseWriter, r *http.Request) {
...
+		httpError(w, fmt.Sprintf("Conflict, The new name %s is already assigned to %s. ", newName, conflictContainer.Id), http.StatusConflict)
```
Comment (aluzzardi) [inline api/handlers.go]: What's the error of `docker` itself when detecting a conflict?

Comment (jimmyxian) [inline api/handlers.go]: Yeah, a little different, i will keep the error info same as docker

File: api/handlers.go (hunk 1 above)
Comment (jimmyxian) [inline api/handlers.go]: I define a Conflict Error when check name in cluster. Where to put better? In cluster/cluster.go Or in cluster/swarm/cluster.go?
Also , i see the flow of create container. it also check name in api/handler.go. Need to fix?

File: cluster/cluster.go (hunk 2 above)
Comment (jimmyxian) [inline cluster/cluster.go]: OK

File: test/integration/api.bats (hunk 3 above)
Comment (jimmyxian) [inline test/integration/api.bats]: OK

**Comment by jimmyxian** [comment]
@aluzzardi 
Modified by the comments

File: cluster/cluster.go [hunk 5]
```diff
@@ -43,4 +43,7 @@ type Cluster interface {
 	// FIXME: remove this method
 	// Return a random engine
 	RANDOMENGINE() (*Engine, error)
+
+	// RenameContainer rename a container
+	RenameContainer(container *Container, newName string) error
```
Comment (vieux) [inline cluster/cluster.go]: small nit, can you leave `RANDOMENGINE() (*Engine, error)` at the bottom ?

**Comment by aluzzardi** [comment]
LGTM

File: cluster/engine.go [hunk 6]
```diff
@@ -555,3 +555,16 @@ func (e *Engine) cleanupContainers() {
...
+// RenameContainer rename a container
+func (e *Engine) RenameContainer(container *Container, newName string) error {
+	// send rename request
+	err := e.client.RenameContainer(container.Id, newName)
+	if err != nil {
+		return err
+	}
+
+	// refresh container
+	err = e.refreshContainer(container.Id, true)
```
Comment (vieux) [inline cluster/engine.go]: @jimmyxian can you make a PR to make this one line ?

File: cluster/swarm/cluster.go [hunk 7]
```diff
@@ -333,3 +333,29 @@ func (c *Cluster) RANDOMENGINE() (*cluster.Engine, error) {
...
+		return err
+	}
+
+	// update container name in store
+	st, err := c.store.Get(container.Id)
+	if err != nil {
+		return err
+	}
+	st.Name = newName
+	err = c.store.Replace(container.Id, st)
```
Comment (vieux) [inline cluster/swarm/cluster.go]: @jimmyxian can you make a PR to make this one line ?

Comment (jimmyxian) [inline cluster/swarm/cluster.go]: OK

**Pull request merged**